onnxruntime/core/framework/tensorprotoutils.cc

```diff
@@ -4,213 +4,577 @@
+        (*out)->rel_path = ret;
+#endif
+      } else if (stringmap.key() == "offset" && !stringmap.value().empty()) {
+        char* end;
+        (*out)->offset = OrtStrtoPtrDiff(stringmap.value().c_str(), &end, 10);
+        if (end != stringmap.value().c_str() + stringmap.value().length())
+          return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "parsing ", stringmap.value(), " failed");
+      } else if (stringmap.key() == "length" && !stringmap.value().empty()) {
+        char* end;
+        (*out)->length = OrtStrtoPtrDiff(stringmap.value().c_str(), &end, 10);
+        if (end != stringmap.value().c_str() + stringmap.value().length())
+          return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "parsing ", stringmap.value(), " failed");
+      } else if (stringmap.key() == "checksum" && !stringmap.value().empty()) {
+        (*out)->checksum = stringmap.value();
+      } else {
+        return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "model format error!");
+      }
+    }
+    if ((*out)->rel_path.empty()) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "model format error! Missing 'location'");
+    }
+    return Status::OK();
+  }
+};
+
+class OrtBuffer {
+ public:
+  virtual const void* GetData() = 0;
+  virtual size_t GetLength() = 0;
+  virtual ~OrtBuffer() = default;
+  ORT_DISALLOW_ASSIGNMENT(OrtBuffer);
+};
+
+#if 0
+class OrtHeapBuffer {
+ public:
+  const void * GetData() {
+
+  }
+  size_t GetLength() {
+
+  }
+  static OrtHeapBuffer* Create(){
+
+  }
+ private:
+  void* data_;
+  size_t length_;
+};
+#endif
+
+#ifdef __GNUC__
+constexpr inline bool IsLittleEndianOrder() noexcept { return __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__; }
+#else
+// On Windows and Mac, this function should always return true
+GSL_SUPPRESS(type .1)  // allow use of reinterpret_cast for this special case
+inline bool IsLittleEndianOrder() noexcept {
+  static int n = 1;
+  return (*reinterpret_cast<char*>(&n) == 1);
+}
+#endif
+
 std::vector<int64_t> GetTensorShapeFromTensorProto(const ONNX_NAMESPACE::TensorProto& tensor_proto) {
   const auto& dims = tensor_proto.dims();
-  std::vector<int64_t> tensor_shape_vec(dims.size());
+  std::vector<int64_t> tensor_shape_vec(static_cast<size_t>(dims.size()));
   for (int i = 0; i < dims.size(); ++i) {
     tensor_shape_vec[i] = dims[i];
   }
 
   return tensor_shape_vec;
 }
 
-std::vector<int64_t> GetTensorShapeFromTensorShapeProto(const ONNX_NAMESPACE::TensorShapeProto& tensor_shape_proto) {
-  const auto& dims = tensor_shape_proto.dim();
-  std::vector<int64_t> tensor_shape_vec(dims.size());
-  for (int i = 0; i < dims.size(); ++i) {
-    tensor_shape_vec[i] = dims[i].has_dim_param()
-                              ? -1 /* symbolic dimensions are represented as -1 in onnxruntime*/
-                              : dims[i].dim_value();
+// This function doesn't support string tensors
+template <typename T>
+static Status UnpackTensorWithRawData(const void* raw_data, size_t raw_data_length, size_t expected_size,
```

Comment:
>UnpackTensorWithRawData [](start = 14, length = 23)

Were there any changes to the existing methods from tensorutils when copying them here? If not I'll skip reviewing them.

Reply:
I changed the implementation of IsLittleEndianOrder function. No others.

---
In reply to: [261416500](https://github.com/Microsoft/onnxruntime/pull/520#discussion_r261416500) [](ancestors = 261416500)